README.md
# enmSdm
Tools for modeling niches and distributions of species

<img align="right" src="enmSdm.png" height="250"/>

<strong>Please note</strong>: `enmSdm` is being replaced by <a href='https://github.com/adamlilith/enmSdmX'>`enmSdmX`</a> (available on <a href='https://cran.r-project.org/web/packages/enmSdmX/index.html'>CRAN</a>).

I am no longer actively developing `enmSdm` because it relies heavily on the soon-to-be-defunct `sp` and (maybe) `raster` packages, which in turn rely on the soon-to-be-defunct `PROJ4` geodetic/coordinate reference system libraries. Its replacement, <a href='https://github.com/adamlilith/enmSdmX'>`enmSdmX`</a> works with the new `terra` and `sf` packages, and--most importantly--`PROJ6`!

This package is a complement to the popular `dismo` package for R by Robert Hijmans. Its contains a suite of efficiency functions for preparing data, training and evaluating species distribution models and ecological niche models, and comparing ecological niches.

You can install this package in R using these commands:

`remotes::install_github('adamlilith/enmSdm', dependencies=TRUE)`  

Note: If for some reason these commands don't work, you can install the package(s) by downloading the latest zip/tar file from the `zipTarFiles` directory and installing the package(s) manually. If you do this, you will also have to install the `legendary` package, which is on GitHub under my account (`adamlilith`), as well as the `omnibus` and `statisfactory` packages on CRAN.

### Data preparation ###
* `elimCellDups`: Eliminate duplicate points in each cell of a raster
* `geoFold`: Generate geographically distinct k-folds
* `geoThin` and `geoThinApprox`: Geographically thin points

### Model training ###
* `trainByCrossValid`: Wrapper for implementing some `trainXYZ` function across cross-validation folds (see also `summaryByCrossValid`).
* `trainBrt`: Boosted regression trees (BRTs)
* `trainCrf`: Conditional regression trees (CRFs)
* `trainGam`: Generalized additive models (GAMs)
* `trainGlm`: Generalized linear models (GLMs)
* `trainGlmDredge`: Generalized linear models (GLMs)
* `trainMaxEnt` and `trainMaxNet`: Maxent models
* `trainNs`: Natural splines (NSs)
* `trainRf`: Random forests (RFs)  

### Model prediction ###
* `predictEnmSdm` Predict most model types using default settings
* `predictMaxNet` Predict MaxNet (MaxEnt) model

### Model evaluation ###
* `aucWeighted`: AUC (with/out site weights)
* `aucMultiWeighted`: Multivariate version of AUC (with/out site weight)
* `contBoyce`: Continuous Boyce Index (with/out site weights)
* `contBoyce2x`: "2X coverage" version of the Continuous Boyce Index (with/out site weights)
* `fpb`: Fpb measure of model discrimination capacity (with/out site weights)
* `thresholdWeighted`: Thresholds to convert continuous predictions to binary predictions (with/out site weights)
* `thresholdStats`: Model performance statistics based on thresholded predictions (with/out site weights)
* `tjursR2Weighted`: Tjur's R2 (with/out site weights)
* `tssWeighted`: True Skill Statistic (TSS) (with/out site weights)
* `modelSize`: Number of response values in a model object

### Niche overlap ###
* `compareNiches`: Niche overlap metrics
* `compareResponse`: Compare niche model responses to a single variable
* `mop`: Calculate mobility-oriented parity, a measure of multivariate distance as per Saupe et al. 2012.
* `nicheOverlap`: Calculate niche overlap as per Broennimann et al. Global Ecology and Biogeography 21:481-497
* `randPointsRespectingSelf`: Randomize geographic points while approximately respecting observed spatial autocorrelation structure between points
* `randPointsRespectingSelfOther2`: Randomize two sets of geographic points while approximately respecting observed spatial autocorrelation structure between and within sets
* `randPointsBatch`: Call `randPointsRespectingSelf` or `randPointsRespectingSelfOther2` multiple times
* `randPointsBatchExtract`: Extract environment from a set of rasters for sets of randomized points generated using `randPointsBatch`
* `randPointsBatchSampled`: Collate all sets of randomized points generated using `randPointsBatch`
* `randPointsBatchNicheOverlap`: Calculate niche overlap between sets of randomized points that were generated using `randPointsBatch`

### Spatial autocorrelation ###
* `spatialCorrForPoints`: Calculate pairwise distance-based measure of global spatial autocorrelation between geographic points
* `spatialCorrForPointsSummary`: Characteristic cluster size of spatial points (distance of global autocorrelation)
* `spatialCorrForPointsPlot`: Plot observed and null distributions of pairwise distance-based measure of global spatial autocorrelation
* `spatialCorrForPointsWeight`: Assign weights to points based on pairwise distance-based measure of global spatial autocorrelation

### Functions for rasters ###
* `bioticVelocity`: Velocity of movement across a series of rasters
* `getCores`: Get number of processor cores
* `extractLarge`: Like `extract()`, but for large-memory spatial objects
* `interpolateRasters`: Interpolate a stack of rasters
* `longLatRasters`: Generate rasters with values of longitude/latitude for cell values
* `rastWithSquareCells`: Create a raster with square cells from an object with an extent
* `sampleRast` and `sampleRastStrat`: Sample raster with/out replacement and possibly in a stratified manner
* `squareRastCells`: Resample a raster so cells are square

### Range area based on minimum convex polygons ###
* `mcpFromPolys`: Minimum convex polygon from a set of polygons *and* points

### Geographic utility functions ###
* `convertTropicosCoords`: Convert coordinates from the TROPICOS database
* `coordPrecision`: Calculate maximum possible coordinate precision
* `createPlotPoly`: Create a `SpatialPolygon` the same size as a plot region
* `decimalToDms`: Convert decimal coordinate to degrees-minutes-seconds
* `dmsToDecimal`: Convert degrees-minutes-seconds coordinate to decimal
* `getCRS`: Return a proj4string (coordinate reference system string) using a nickname
* `getEPSG`: Return a EPSG code (coordinate reference system codeg) using a nickname
* `pointDist`: Geographic distance between set(s) of points
* `svToSpatial`: Convert SpatVector object to a Spatial* object
* `xToCoords`: Extract geographic coordinates from a data frame, matrix, or SpatialPoints* object

### Data
* `lemurs`: Lemur occurrences
* `mad0`: Madagascar spatial object

## Citation ##
As of October 2020 there is no package-specific publication for `enmSdm`, but the package was first used and cited in:

Morelli*, T.L., Smith*, A.B., Mancini, A.N., Balko, E. A., Borgenson, C., Dolch, R., Farris, Z., Federman, S., Golden, C.D., Holmes, S., Irwin, M., Jacobs, R.L., Johnson, S., King, T., Lehman, S., Louis, E.E. Jr., Murphy, A., Randriahaingo, H.N.T., Lucien, Randriannarimanana, H.L.L., Ratsimbazafy, J., Razafindratsima, O.H., and Baden, A.L. 2020. The fate of Madagascar’s rainforest habitat.  **Nature Climate Change** 10:89-96. * Equal contribution https://doi.org/10.1038/s41558-019-0647-x

**Abstract**. Madagascar has experienced extensive deforestation and overharvesting, and anthropogenic climate change will compound these pressures. Anticipating these threats to endangered species and their ecosystems requires considering both climate change and habitat loss effects. The genus **Varecia** (ruffed lemurs), which is composed of two Critically Endangered forest-obligate species, can serve as a status indicator of the biodiverse eastern rainforest of Madagascar. Here, we combined decades of research to show that the suitable habitat for ruffed lemurs could be reduced by 29–59% from deforestation, 14–75% from climate change (representative concentration pathway 8.5) or 38–93% from both by 2070. If current protected areas avoid further deforestation, climate change will still reduce the suitable habitat by 62% (range: 38–83%). If ongoing deforestation continues, the suitable habitat will decline by 81% (range: 66–93%). Maintaining and enhancing the integrity of protected areas, where rates of forest loss are lower, will be essential for ensuring persistence of the diversity of the rapidly diminishing Malagasy rainforests.


NEWS.md
enmSdm X.X.X 2022-
Added parallelization and memory-friendly feature
to predictEnmSdm!

enmSdm 0.9.3 2022-05-12
trainMaxNet() can now multi-core!

enmSdm 0.9.2 2022-04-17
extractLarge() is preserves memory much better

enmSdm 0.9.1 2022-04-15
Added function extractLarge()
rastWithSquareCells() and squareRastCells() now terra-compatible
Fixed bug in predictMaxNet()

enmSdm 0.9.0 2022-03-15
Added function predictMaxNet()

enmSdm 0.8.6 2022-03-15
Fixed bug in predictEnmSdm() when using maxnet()

enmSdm 0.8.5 2022-02-22
Fixed bug in elimCellDups() introduced by fixing
  bug in xToCoords()!

enmSdm 0.8.4 2022-02-22
Fixed bug in other functions introduced by fixing
  bug in xToCoords()!

enmSdm 0.8.3 2022-02-17
Fixed bug in xToCoords()

enmSdm 0.8.2 2022-02-14
Fixed help in trainByCrossValid()

enmSdm 0.7.2 2022-01-29
Fixed bug in sampleRast()

enmSdm 0.7.0 2021-12-15
sampleRast() randomly places sites within cells (not just cell centers)
trainNs() does not call brglmFit2 by default anymore

enmSdm 0.6.1 2021-12-10
Added createPlotPoly()

enmSdm 0.6.0 2021-11-10
Fixing R CMD check issues

enmSdm 0.5.3.8 2021-10-25
Update for Maxent 3.4.4

enmSdm 0.5.3.6 2021-06-22
Fixed mistake in how Pearson correlation is named in
compareNiches() and compareResponse() (was "rho", is now "cor")

enmSdm 0.5.3.5 2021-04-14
Renamed aucMultiWeighted() cases to be more intuitive

enmSdm 0.5.3.5 2021-04-14
Added thresholdMultiWeighted()
Better handling of extreme cases in aucMultiWeighted()

enmSdm 0.5.3.4 2021-04-12
Fixed bug in aucMultiWeighted()

enmSdm 0.5.3.3 2021-03-22
Add getEPSG() for returning EPSG codes

enmSdm 0.5.3.2 2021-02-05
Fixed bug in compareResponse() for areal calculations

enmSdm 0.5.3.1 2021-02-01
Fixed runtime bug in bioticVelocity()

enmSdm 0.5.3.0 2020-11-24
Fixed error in compareNiches() and bioticVelocity() related to calculating
Hellinger's I (noted in Erratum to Warren et al. 2008--thank you, João Carlos!)

enmSdm 0.5.2.9 2020-10-20
bioticVelocity() better estimates quantile velocities.
bioticVelocity() speed improvements, including multi-core.

enmSdm 0.5.2.8 2020-10-12
trainGlm() now catches cases when no viable models are returned.

enmSdm 0.5.2.7 2020-10-05
Added new methods and a fail-safe to interpolateRasters()

enmSdm 0.5.2.6 2020-10-03
Updated bioticVelocity() to use rasters, rather than arrays

enmSdm 0.5.2.4 2020-09-29
Fixed bug in interpolateRasters() when only one output raster is being interpolated

enmSdm 0.5.2.3 2020-09-24
Cell-weighted calculation of quantile lat/long in bioticVelocity()

enmSdm 0.5.2.3 2020-09-23
Remove ability bioticVelocity() to use "pophist" objects
New, straightforward examples in bioticVelocity()
Fixed "quant" and "Quants" velocities in bioticVelocity()

enmSdm 0.5.2.2
Graceful catch of non-converged/insufficient models in trainBrt()
Sensical names in output of bioticVelocity()

enmSdm 0.5.2.1
Consistent output across models for summaryByCrossValid()
Clarified help for bioticVelocity()

enmSdm 0.5.2.0
trainNs() can now return a list of all models
trainNs() now cycles through degrees of freedom
Can now use natural splines (trainNS()) with cross validation functions
Fixed issues with trainByCrossValid()

enmSdm 0.5.1.9
enmSdm 0.5.1.8
Create rastWithSquareCells()
Fix squareRasterCells(); renamed to squareRastCells()
Remove export of localSpatialCorrForValues() because of instability issue

enmSdm 0.5.1.7
Add squareRasterCells()

enmSdm 0.5.1.6
coordPrecision() handles coordinates in DMS format

enmSdm 0.5.1.5
Add Robinson projection to getCRS()

enmSdm 0.5.1.3
Recompiled for R 4.0.0

enmSdm 0.5.1.2
trainBrt() parallelized
trainMaxEnt() parallelized

enmSdm 0.5.0.4
trainByCrossValid() handles non-converged models

enmSdm 0.5.0.3
bioticVelocity() can correct for non-shared land mass
Fix bug in getCRS()

enmSdm 0.5.0.2
Fix bug in bioticVelocity()

enmSdm 0.5.0.1
Fix bug in spatialCorrForValues()

enmSdm 0.5.0.0
Fix bug in predictMaxEnt()

enmSdm 0.4.0.5
trainMaxEnt() and trainMaxNet() internals harmonized
Made sensible examples for trainXYZ() functions

enmSdm 0.4.0.3
Added function makeCRS() for making custom coordinate reference systems

enmSdm 0.4.0.2
Added function decimalToDms() for converting decimal coordinates to degrees-minutes-seconds format
Fixed bug in spatialCorrForPoints() when supplying distance bins in km

enmSdm 0.4.0.1
Fixed bug in coordPrecision() when supplied value was NA.

enmSdm 0.4.0.0
Fixed bug in elimCellDups() when there were no cell duplicates.

R/getCRS.r
#' Return a "proj4" string for a datum and projection
#'
#' This function returns the "proj4" string for a particular datum and possibly projection. It is intended to be easier to use than looking up particular proj4 strings (i.e., on the web).
#' @param x Character. Name of proj4 string to return. Possible values include:
#'	\itemize{
#'	\item \code{albersNA} Albers equal-area conic for North America: \code{+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=37.5 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +datum=NAD83 +units=m +no_defs}
#'	\item \code{albersEur} Albers equal-area conic for Europe: \code{+proj=aea +lat_1=43 +lat_2=62 +lat_0=30 +lon_0=10 +x_0=0 +y_0=0 +ellps=intl +units=m +no_defs}
#'	\item \code{madAlbers} Albers equal-area conic for Madagascar: \code{+proj=aea +lat_1=-14 +lat_2=-24 +lat_0=-19 +lon_0=47 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs +ellps=WGS84}
#'	\item \code{climateNA} ClimateNA dataset: \code{+proj=lcc +lat_1=49 +lat_2=77 +lat_0=0 +lon_0=-95 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs +ellps=WGS84}
#'	\item \code{chelsa} CHELSA climate dataset (actually WGS84): or \code{+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84}
#'	\item \code{dayMet} DayMet dataset: \code{+proj=lcc +lon_0=-100 +lat_0=42.5 +x_0=0 +y_0=0 +a=6378137 +rf=298.257223563 +lat_1=25 +lat_2=60}
#'	\item \code{lccCONUS} Lambert conformal conic for the contiguous United States: \code{+proj=lcc +lat_1=33 +lat_2=45 +lat_0=39 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +datum=NAD83 +units=m +no_defs}
#'	\item \code{lccEp} Lambert conformal conic for Europe: \code{+proj=lcc +lat_1=43 +lat_2=62 +lat_0=30 +lon_0=10 +x_0=0 +y_0=0 +ellps=intl +units=m +no_defs}
#'	\item \code{lccNA} Lambert conformal conic for North America: \code{+proj=lcc +lat_1=20 +lat_2=60 +lat_0=40 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +datum=NAD83 +units=m +no_defs}
#'	\item \code{lccNAsia} Lambert conformal conic for North Asia: \code{+proj=lcc +lat_1=15 +lat_2=65 +lat_0=30 +lon_0=95 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs}
#'	\item \code{lccSA} Lambert conformal conic for South America: \code{+proj=lcc +lat_1=-5 +lat_2=-42 +lat_0=-32 +lon_0=-60 +x_0=0 +y_0=0 +ellps=aust_SA +units=m +no_defs}
#'	\item \code{lccSAsia} Lambert conformal conic for South Asia: \code{+proj=lcc +lat_1=7 +lat_2=-32 +lat_0=-15 +lon_0=125 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs}
#'	\item \code{laeaN} Lambert azimuthal equal area for the North Pole: \code{+proj=laea +lat_0=90 +lon_0=0 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs}
#'	\item \code{laeaS} Lambert azimuthal equal area for the South Pole: \code{+proj=laea +lat_0=-90 +lon_0=0 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs}
#'	\item \code{mollweide} Mollweide equal-area for the world: \code{+proj=moll +lon_0=0 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs}
#'	\item code{nad27} NAD27 (unprojected): \code{+proj=longlat +datum=NAD27 +no_defs +ellps=clrk66 +nadgrids=@conus,@alaska,@ntv2_0.gsb,@ntv1_can.dat}
#'	\item \code{nad83} NAD83 (unprojected): \code{+proj=longlat +datum=NAD83 +no_defs +ellps=GRS80}
#'	\item \code{nad83harn} NAD83 HARN (unprojected): \code{+proj=longlat +ellps=GRS80 +no_defs}
#'	\item \code{plateCaree} Plate Caree using WGS84: \code{+proj=eqc +lat_ts=0 +lat_0=0 +lon_0=0 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_def}
#'	\item \code{prism} PRISM dataset (actually WGS84): \code{+proj=longlat +ellps=GRS80,0,0,0,0 +no_defs}
#'	\item \code{robinson} Robinson: \code{+proj=robin +lon_0=0 +x_0=0 +y_0=0}
#'	\item \code{wgs72} CRS for WGS72 (unprojected): \code{+proj=longlat +ellps=WGS72 +no_defs}
#'	\item \code{wgs84} CRS for WGS84 (unprojected): \code{+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84}
#'	\item \code{worldclim} WORLDCLIM climate data set (actually WGS84): \code{+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84}
#'  \item UTM projections require the datum (e.g., NAD27, NAD83, WGS84, etc.), the hemisphere (north or south) and the zone. The "code" is as:
#' \code{utm<coordinate system><hemisphere><zone>}. For example, \code{utmNad83north9} returns the CRS for UTM zone 9 in the northern hemisphere using the NAD83 coordinate reference system. Likewise, \code{utmNad27south6} returns the CRS for UTM zone 6 in the southern hemisphere using the NAD27 coordinate system. So far the following have been implemented:
#' 	\itemize{
#' 		\item Coordinate Systems: \code{nad27}, \code{nad83}, \code{wgs84}
#' 		\item Hemispheres: \code{north}
#' 	}
#' }
#' @param asCRS Logical. If \code{TRUE} then return object is of class \code{CRS}. If \code{FALSE} (default) then returned object is of class \code{character}.
#' @return Object of class \code{CRS} or \code{character}.
#' @seealso \code{\link[birdsEye]{makeCRS}}
#' @examples
#' getCRS('wgs84')
#' getCRS('prism')
#' getCRS('mollweide', TRUE)
#' @export

getCRS <- function(
	x,
	asCRS = FALSE
) {

	x <- tolower(x)

	out <- if (x == 'albersna') {
		'+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=37.5 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +datum=NAD83 +units=m +no_defs'
	} else if (x == 'alberseur') {
		'+proj=aea +lat_1=43 +lat_2=62 +lat_0=30 +lon_0=10 +x_0=0 +y_0=0 +ellps=intl +units=m +no_defs'
	} else if (x == 'chelsa') {
		'+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84'
	} else if (x == 'climatena') {
		'+proj=lcc +lat_1=49 +lat_2=77 +lat_0=0 +lon_0=-95 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs +ellps=WGS84'
	} else if (x == 'daymet') {
		'+proj=lcc +lon_0=-100 +lat_0=42.5 +x_0=0 +y_0=0 +a=6378137 +rf=298.257223563 +lat_1=25 +lat_2=60'
	} else if (x == 'lccnasia') {
		'+proj=lcc +lat_1=15 +lat_2=65 +lat_0=30 +lon_0=95 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs'
	} else if (x == 'lccsasia') {
		'+proj=lcc +lat_1=7 +lat_2=-32 +lat_0=-15 +lon_0=125 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs'
	} else if (x == 'lccsa') {
		'+proj=lcc +lat_1=-5 +lat_2=-42 +lat_0=-32 +lon_0=-60 +x_0=0 +y_0=0 +ellps=aust_SA +units=m +no_defs'
	} else if (x == 'lccep') {
		'+proj=lcc +lat_1=43 +lat_2=62 +lat_0=30 +lon_0=10 +x_0=0 +y_0=0 +ellps=intl +units=m +no_defs'
	} else if (x == 'lccna') {
		'+proj=lcc +lat_1=20 +lat_2=60 +lat_0=40 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +datum=NAD83 +units=m +no_defs'
	} else if (x == 'lccconus') {
		'+proj=lcc +lat_1=33 +lat_2=45 +lat_0=39 +lon_0=-96 +x_0=0 +y_0=0 +ellps=GRS80 +datum=NAD83 +units=m +no_defs'
	} else if (x == 'laean') {
		'+proj=laea +lat_0=90 +lon_0=0 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs'
	} else if (x == 'laeas') {
		'+proj=laea +lat_0=-90 +lon_0=0 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs'
	} else if (x == 'madalbers') {
		'+proj=aea +lat_1=-14 +lat_2=-24 +lat_0=-19 +lon_0=47 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs +ellps=WGS84'
	} else if (x == 'mollweide') {
		'+proj=moll +lon_0=0 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs'
	} else if (x == 'nad27') {
		'+proj=longlat +datum=NAD27 +no_defs +ellps=clrk66 +nadgrids=@conus,@alaska,@ntv2_0.gsb,@ntv1_can.dat'
	} else if (x == 'nad83') {
		'+proj=longlat +datum=NAD83 +no_defs +ellps=GRS80'
	} else if (x == 'nad83harn') {
		'+proj=longlat +ellps=GRS80 +no_defs'
	} else if (x == 'platecaree') {
		'+proj=eqc +lat_ts=0 +lat_0=0 +lon_0=0 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs +ellps=WGS84'
	} else if (x == 'prism') {
		'+proj=longlat +ellps=GRS80,0,0,0,0 +no_defs'
	} else if (x == 'robinson') {
		'+proj=robin +lon_0=0 +x_0=0 +y_0=0'
	} else if (x == 'wgs72') {
		'+proj=longlat +ellps=WGS72 +no_defs'
	} else if (x == 'wgs84') {
		'+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84'
	} else if (x == 'worldclim') {
		'+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84'
	} else if (substr(x, 1, 13) == 'utmnad27north') {
		paste0('+proj=utm +zone=', substr(x, nchar(x) - 1, nchar(x)), ' +datum=NAD27 +units=m +no_defs +ellps=clrk66 +nadgrids=@conus,@alaska,@ntv2_0.gsb,@ntv1_can.dat')
	} else if (substr(x, 1, 13) == 'utmnad83north') {
		paste0('+proj=utm +zone=',  substr(x, nchar(x) - 1, nchar(x)), ' +datum=NAD83 +units=m +no_defs +ellps=GRS80')
	} else if (substr(x, 1, 13) == 'utmwgs84north') {
		paste0(' +proj=utm +zone=', substr(x, nchar(x) - 1, nchar(x)), ' +datum=WGS84 +units=m +no_defs +ellps=WGS84')
	} else {
		NA
	}

	if (asCRS) out <- sp::CRS(out)
	out

}
